Disable cert validation when monkeypatching AppEngine

bendemaree opened the pull request:
We have a use case for this in which, due to an issue with AppEngine's `dev_appserver.py`, certain certificates will fail validation; more on that issue [here](https://code.google.com/p/googleappengine/issues/detail?id=13154). The only real workaround when using the emulator is to disable certificate validation.

Because this is a large code base, in which requests is used both explicitly and within third-party libraries, we are using the monkeypatch; however, to my eye there's no way to set the `validate_certificate` kwarg in the constructor for `AppEngineAdapter` to `True` when monkeypatching; if we could, it would serve our use case by allowing global disabling of certificate validation when a local environment is detected.

This creates a new `AppEngineInsecureAdapter` in which cert validation is disabled by default and adds a kwarg to `monkeypatch()` to trigger its use. Initially I tried to simply make the monkeypatch set `partial(AppEngineAdapter, validate_certificate=False)` but that was a bit naive as the adapter is not just instantiated.

sigmavirus24 commented on requests_toolbelt/adapters/appengine.py:
```diff
@@ -59,6 +78,22 @@ def init_poolmanager(self, connections, maxsize, block=False):
+    """A variant of the the transport adapter for Requests to use urllib3's GAE
+    support that does not validate certificates by default. Use with caution!
+
+    See `.AppEngineAdapter` for further details.
```
```
`.AppEngineAdapter`
```

Isn't valid ReStructuredText it needs to be

```rst
:class:`AppEngineAdapter`
```

sigmavirus24 commented on requests_toolbelt/adapters/appengine.py:
```diff
@@ -59,6 +78,22 @@ def init_poolmanager(self, connections, maxsize, block=False):
         self.poolmanager = _AppEnginePoolManager(self._validate_certificate)
 
 
+class AppEngineInsecureAdapter(adapters.HTTPAdapter):
```
Better question, why did you copy an entire class definition instead of doing

```py
class InsecureAppEngineAdapter(AppEngineAdapter):

    def __init__(self, validate_certificate=False, *args, **kwargs):
        super(InsecureAppEngineAdapter, self).__init__(validate_certificate, *args, **kwargs)
```

The AppEngineAdapter already creates the `_AppEnginePoolManager`.

sigmavirus24 replied:
No worries.

sigmavirus24 commented:
> Initially I tried to simply make the monkeypatch set partial(AppEngineAdapter, validate_certificate=False) but that was a bit naive as the adapter is not just instantiated.

Can you explain this a bit further? What problems did you find with this approach?

bendemaree commented:
I ran into [this check](https://github.com/kennethreitz/requests/blob/master/requests/adapters.py#L117) which resulted in

```
  File "requests_toolbelt/adapters/appengine.py", line 76, in __init__
    super(AppEngineAdapter, self).__init__(*args, **kwargs)
  File "equests/adapters.py", line 117, in __init__
    super(HTTPAdapter, self).__init__()
TypeError: super() argument 1 must be type, not functools.partial
```

sigmavirus24 commented:
I suspect the monkey-patch is patching the wrong thing then (or we're mis-using imports in the toolbelt).

sigmavirus24 commented:
What happens if you delete [this line](https://github.com/sigmavirus24/requests-toolbelt/blob/7679220dc272e0c824eed8fa12d6adc702e21bcb/requests_toolbelt/adapters/appengine.py#L138)?

bendemaree commented:
I'll give it a go in a few hours here. The possibility of the patching being off might explain some confusing errors I got in my initial attempt to write some test around this as well. Thanks for the pointers!

sigmavirus24 commented:
@bendemaree excellent investigation!

So looking into #165 again, I have mixed feelings about it, but it seems to be completely valid. So I guess we're back to the method that you're currently employing in this pull request (having two different adapter).

I think this is best done like so:

```py
class InsecureAppEngineAdapter(AppEngineAdapter):

    def __init__(self, *args, **kwargs):  # Note: there's no validate_certificate parameter here
        # Maybe ensure "validate_certificate" is not in **kwargs
        super(InsecureAppEngineAdapter, self).__init__(validate_certificate=False, *args, **kwargs)
```

And then we should key the monkey-patch as you're currently doing. Thanks for sticking with this @bendemaree

bendemaree commented:
@sigmavirus24 My pleasure, thanks for all you do!

I added some tests around this, and the monkeypatch itself, now that everything is making sense. I also added a warning if the user instantiates `InsecureAppEngineAdapter` manually with `validate_certificate` specified (which I'm hoping will never happen!). I went on the precedent of `warnings` already being in use; however, I copy-pasted from `adapters/socket_options.py` and I think [the usage there](https://github.com/sigmavirus24/requests-toolbelt/blob/master/requests_toolbelt/adapters/socket_options.py#L46) might be incorrect (message then class unless first arg is an instance).

Lastly--I took a hint from your notes and renamed it to `InsecureAppEngineAdapter`...I had a suspicion that was the more intuitive name myself.

sigmavirus24 commented on docs/adapters.rst:
```diff
@@ -53,6 +53,25 @@ ways to take advantage of this support at the moment:
+       will *not* validate certificates. This effectively sets the
+       ``validate_certificate`` argument to urlfetch.Fetch() to ``False``. You
+       should avoid using this wherever possible. Details can be found in the
+       `documentation for urlfetch.Fetch()`__.
```
This should end in a single underscore

sigmavirus24 commented on docs/adapters.rst:
```diff
@@ -53,6 +53,25 @@ ways to take advantage of this support at the moment:
+       should avoid using this wherever possible. Details can be found in the
+       `documentation for urlfetch.Fetch()`__.
+
+       __ https://cloud.google.com/appengine/docs/python/refdocs/google.appengine.api.urlfetch
```
This should be

```rst
.. _documentation for urlfetch.Fetch(): <url>
```

bendemaree replied:
It's almost like I'm just winging the RST... 😬

sigmavirus24 commented on docs/adapters.rst:
```diff
@@ -53,6 +53,25 @@ ways to take advantage of this support at the moment:
 
        appengine.monkeypatch()
 
+If you should need to disable certificate validation when monkeypatching (to
```
Perhaps we should add a link ref here, e.g.,

```
.. _insecure_appengine:

If you should need to disable ...
```

sigmavirus24 commented on requests_toolbelt/adapters/appengine.py:
```diff
@@ -29,8 +29,28 @@
+       should avoid using this wherever possible. Details can be found in the
+       `documentation for urlfetch.Fetch()`__.
+
+       __ https://cloud.google.com/appengine/docs/python/refdocs/google.appengine.api.urlfetch  # noqa
```
Same thing as above, although this looks duplicated from there. Perhaps remove the duplication and add a reference to it, e.g.,

```rst
You are also able to :ref:`disable certificate validation <insecure_appengine>` when monkey-patching.
```

sigmavirus24 commented on requests_toolbelt/adapters/appengine.py:
```diff
@@ -59,6 +79,24 @@ def init_poolmanager(self, connections, maxsize, block=False):
+    """A variant of the the transport adapter for Requests to use urllib3's GAE
+    support that does not validate certificates by default. Use with caution!
+
+    See :class:`AppEngineAdapter` for further details.
```
We should probably add to the docstring:

```rst

.. note::

    The ``validate_certificate`` keyword argument will not be honored here and is not part of the signature because we always force it to ``False``.
```

sigmavirus24 commented on requests_toolbelt/adapters/appengine.py:
```diff
@@ -59,6 +79,24 @@ def init_poolmanager(self, connections, maxsize, block=False):
 
 
+class InsecureAppEngineAdapter(AppEngineAdapter):
+    """A variant of the the transport adapter for Requests to use urllib3's GAE
```
I know we don't make pydocstyle a required check in Travis, but please try to follow PEP-0257. For example,

```
An always-insecure GAE adapter for Requests.

This is a variant of the AppEngineAdapter provided by this module that will **never** validate certificates by default. **Use with caution!**
```

bendemaree replied:
Whoops, sloppy; thanks for the catch. Fixed in the warning docstring as well.

sigmavirus24 commented on requests_toolbelt/adapters/appengine.py:
```diff
@@ -123,19 +161,29 @@ def urlopen(self, method, url, body=None, headers=None, retries=None,
     # instead of overriding an imported HTTPAdapter as is done here.
-    sessions.HTTPAdapter = AppEngineAdapter
-    adapters.HTTPAdapter = AppEngineAdapter
+    if validate_certificate:
```
Since there's other review feedback here, what if we do:

```py
adapter = AppEngineAdapter
if not validate_certificate:
    adapter = InsecureAppEngineAdapter
```

sigmavirus24 commented on requests_toolbelt/exceptions.py:
```diff
@@ -23,3 +23,12 @@ class RequestsVersionTooOld(Warning):
     pass
+
+
+class IgnoringCertificateValidation(Warning):
```
Should probably make this mildly more specific, e.g.,

```
class IgnoringGAECertificateValidation(Warning):
```

sigmavirus24 commented on requests_toolbelt/exceptions.py:
```diff
@@ -17,17 +17,20 @@ class VersionMismatchError(Exception):
 
 class RequestsVersionTooOld(Warning):
-    """Used to indiciate that the Requests version is too old.
+    """Used to indicate that the Requests version is too old.
```
Nice catch!

sigmavirus24 commented:
This looks excellent! Thank you!

bendemaree commented:
Rock on, thanks for dealing with my poor doc skills. And, stay warm down there! ❄️

sigmavirus24 commented:
> And, stay warm down there

I'll do my best!

> thanks for dealing with my poor doc skills

Your first attempts are 100x better than what I used to write. Documentation writing is something I struggle with constantly. Most of my edits were syntax related (which is the easiest part of ReStructuredText for me).

bendemaree commented:
@sigmavirus24 Thanks for the kind words!

Looks like there's a little more work to do here...I went to plugin this in to the project and got more infinite recursion errors. I then installed it into the toy app above; same thing. I went back again to #165 and re-thought things...can you give your reasoning why the change was valid? I don't see any way that `AppEngineAdapter` can inherit from `HTTPAdapater`...and then _replace_ it as `requests.adapters.HTTPAdapter`. As soon as `AppEngineAdapter` is instantiated, it calls super, which, due to the monkeypatch, is itself.

IMO the original request to patch might be invalid, because in that case a specific adapter is being requested. The solution originally posed in the PR seems more valid (importing the adapter assigned to `requests.session`).

sigmavirus24 commented:
So the only reason I think the original report *might* be valid is that Requests doesn't really document *one* way to import an use an HTTPAdapter yourself. So if you look at the code-base that was running into problems with that, they were importing HTTPAdapter from `requests.adapters` which is just as valid as importing it from `requests.sessions`.

Unfortunately @Lukasa is on holiday, so we can't really gather his feedback right now. Let me see if I can find someone else who has opinions about this kind of stuff. Maybe @jonparrott

jonparrott commented:
I'm on my way out as well, I'll see if I can take a closer look later today
before I disappear.

On Fri, Dec 16, 2016, 10:48 AM Ian Cordasco <notifications@github.com>
wrote:

> So the only reason I think the original report *might* be valid is that
> Requests doesn't really document *one* way to import an use an
> HTTPAdapter yourself. So if you look at the code-base that was running into
> problems with that, they were importing HTTPAdapter from requests.adapters
> which is just as valid as importing it from requests.sessions.
>
> Unfortunately @Lukasa <https://github.com/Lukasa> is on holiday, so we
> can't really gather his feedback right now. Let me see if I can find
> someone else who has opinions about this kind of stuff. Maybe @jonparrott
> <https://github.com/jonparrott>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/sigmavirus24/requests-toolbelt/pull/169#issuecomment-267667004>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAPUc3NWMj9e-k6Rm-Gw0Jg6nv9U4B_Sks5rItzDgaJpZM4LJU6N>
> .
>

sigmavirus24 commented:
Thanks @jonparrott

jonparrott commented:
> So the only reason I think the original report might be valid is that Requests doesn't really document one way to import an use an HTTPAdapter yourself. So if you look at the code-base that was running into problems with that, they were importing HTTPAdapter from requests.adapters which is just as valid as importing it from requests.sessions.

Yeah this import aliasing is going to cause all kinds of fun stuff when you try to replace something with its own subclass.

You could do some jiggerypockery where you create a reference to the original `HTTPAdapter` class before patching and ensure that `AppEngineAdapter` inherits from that.

All of that seems pretty gross, though.

sigmavirus24 commented:
> All of that seems pretty gross, though.

I agree 110%. :)

I didn't like the monkey-patching idea at all, but it seemed necessary. Now it seems to be a nightmare. :)

jonparrott commented:
Maybe we should just ask requests to make it easier to do this without
monkey patching.

On Fri, Dec 16, 2016, 12:51 PM Ian Cordasco <notifications@github.com>
wrote:

> All of that seems pretty gross, though.
>
> I agree 110%. :)
>
> I didn't like the monkey-patching idea at all, but it seemed necessary.
> Now it seems to be a nightmare. :)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/sigmavirus24/requests-toolbelt/pull/169#issuecomment-267692857>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAPUc2mE6HOdcL76TVNlEs9W5rz3ONytks5rIvm1gaJpZM4LJU6N>
> .
>

sigmavirus24 commented:
@jonparrott the problem is that requests-toolbelt aims to support as many versions of requests as possible, so that would only work for people who keep up-to-date with the latest and greatest.

bendemaree commented:
@sigmavirus24 @jonparrott have opened a PR with a potential fix; PTAL! 😄
